File: common/secrets.py
"""Hardcoded secrets check — from global/secrets.md.

Flags credential-like literals assigned directly in source code.
Secrets belong in ~/.env/ and must never be copied into project files.
"""

from __future__ import annotations
from pathlib import Path
from typing import Generator
import re

from .issue import Issue, Severity

_RULE = "global/secrets"

# Assignment patterns: key = "value" where key suggests a credential
_SECRET_KEY = re.compile(
    r"""(?ix)
    \b(
        password | passwd | pwd |
        api[_-]?key | apikey |
        secret[_-]?key | client[_-]?secret |
        access[_-]?token | auth[_-]?token | bearer[_-]?token |
        private[_-]?key | signing[_-]?key |
        database[_-]?url | db[_-]?password |
        aws[_-]?secret | aws[_-]?access | aws[_-]?key
    )\s*[=:]\s*["'][^"']{4,}["']
    """
)

# PEM private key headers
_PEM_KEY = re.compile(r"-----BEGIN\s+(RSA|EC|OPENSSH|PRIVATE)\s+PRIVATE KEY-----")

# Skip files that are clearly test fixtures, templates, or documentation
_SKIP_PARTS = {"test", "tests", "fixtures", "examples", "docs", "__pycache__"}
_SKIP_SUFFIXES = {".md", ".txt", ".rst", ".toml", ".example", ".template"}


def check(path: Path, lines: list[str]) -> Generator[Issue, None, None]:
    if path.suffix.lower() in _SKIP_SUFFIXES:
        return
    if any(part in _SKIP_PARTS for part in path.parts):
        return

    for lineno, raw in enumerate(lines, start=1):
        stripped = raw.lstrip()
        if stripped.startswith(("#", "//", "*")):  # skip pure comment lines
            continue

        if m := _SECRET_KEY.search(raw):
            key_name = m.group(1)
            yield Issue(
                file=path, line=lineno, col=m.start() + 1,
                severity=Severity.ERROR,
                rule=_RULE,
                message=(
                    f"hardcoded credential '{key_name}' — "
                    f"move to ~/.env/ and load via environment variable"
                ),
            )

        if m := _PEM_KEY.search(raw):
            yield Issue(
                file=path, line=lineno, col=m.start() + 1,
                severity=Severity.ERROR,
                rule=_RULE,
                message="PEM private key in source file — must never be committed",
            )

File: crates/scanner/src/checks/secrets.rs
use std::path::Path;

use regex::Regex;
use std::sync::LazyLock;

use crate::config::Config;
use crate::context::FileContext;
use crate::issue::{Issue, Severity};

static CREDENTIAL_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?i)\b(password|passwd|pwd|api[-_]?key|apikey|secret[-_]?key|client[-_]?secret|access[-_]?token|auth[-_]?token|bearer[-_]?token|private[-_]?key|signing[-_]?key|database[-_]?url|db[-_]?password|aws[-_]?secret|aws[-_]?access|aws[-_]?key)\s*[=:]\s*["'][^"']{4,}["']"#).unwrap()
});

static PEM_KEY_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"-----BEGIN\s+(RSA|EC|OPENSSH|PRIVATE)\s+PRIVATE KEY-----").unwrap()
});

/// Skip these directories entirely.
const SKIP_DIRS: &[&str] = &["test", "tests", "fixtures", "examples", "docs"];

/// Check for hardcoded secrets and credentials.
pub fn check(
    _file_ctx: &FileContext,
    lines: &[&str],
    _cfg: &Config,
    issues: &mut Vec<Issue>,
    path: &Path,
) {
    let path_str = path.to_string_lossy();
    if SKIP_DIRS.iter().any(|d| path_str.contains(d)) {
        return;
    }

    let mut in_test_module = false;
    let mut test_brace_depth: i32 = 0;

    for (line_num, line) in lines.iter().enumerate() {
        let trimmed = line.trim();

        // Track #[cfg(test)] mod blocks — skip test data
        if trimmed == "#[cfg(test)]" || trimmed == "#[test]" {
            in_test_module = true;
        }
        if in_test_module {
            test_brace_depth += line.chars().filter(|&c| c == '{').count() as i32;
            test_brace_depth -= line.chars().filter(|&c| c == '}').count() as i32;
            if test_brace_depth > 0 || trimmed.starts_with("#[") {
                continue;
            }
            if test_brace_depth <= 0 {
                in_test_module = false;
                test_brace_depth = 0;
                continue;
            }
        }

        if trimmed.starts_with("//") || trimmed.starts_with('#') || trimmed.starts_with('*') {
            continue;
        }

        if CREDENTIAL_RE.is_match(line) {
            issues.push(Issue::new(
                path,
                line_num + 1,
                1,
                Severity::Error,
                "global/secrets",
                "hardcoded credential — use environment variable or secret manager",
            ));
        }

        if PEM_KEY_RE.is_match(line) {
            issues.push(Issue::new(
                path,
                line_num + 1,
                1,
                Severity::Error,
                "global/secrets",
                "embedded private key — never commit keys to source",
            ));
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::context::Language;

    fn make_ctx() -> FileContext {
        FileContext {
            language: Language::Rust,
            is_test_file: false,
            is_mother_file: false,
            is_definition_file: false,
        }
    }

    #[test]
    fn finds_password() {
        let mut issues = Vec::new();
        let lines = vec![r#"let password = "hunter2";"#];
        check(&make_ctx(), &lines, &Config::default(), &mut issues, Path::new("src/main.rs"));
        assert_eq!(issues.len(), 1);
        assert!(issues[0].message.contains("credential"));
    }

    #[test]
    fn finds_pem_key() {
        let mut issues = Vec::new();
        let lines = vec!["-----BEGIN RSA PRIVATE KEY-----"];
        check(&make_ctx(), &lines, &Config::default(), &mut issues, Path::new("src/main.rs"));
        assert_eq!(issues.len(), 1);
        assert!(issues[0].message.contains("private key"));
    }

    #[test]
    fn skips_comment_lines() {
        let mut issues = Vec::new();
        let lines = vec![r#"// password = "test1234""#];
        check(&make_ctx(), &lines, &Config::default(), &mut issues, Path::new("src/main.rs"));
        assert!(issues.is_empty());
    }

    #[test]
    fn skips_test_dirs() {
        let mut issues = Vec::new();
        let lines = vec![r#"let password = "hunter2";"#];
        check(&make_ctx(), &lines, &Config::default(), &mut issues, Path::new("tests/auth.rs"));
        assert!(issues.is_empty());
    }
}
